File: swim-system-java/swim-core-java/swim.uri/src/main/java/swim/uri/UriHostIPv6.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.uri;

import swim.codec.Output;

final class UriHostIPv6 extends UriHost {
  final String address;
  String string;

  UriHostIPv6(String address) {
    this.address = address;
  }

  @Override
  public String address() {
    return this.address;
  }

  @Override
  public String ipv6() {
    return this.address;
  }

  @Override
  public void debug(Output<?> output) {
    output = output.write("UriHost").write('.').write("ipv6").write('(').debug(this.address).write(')');
  }

  @Override
  public void display(Output<?> output) {
    if (this.string != null) {
      output = output.write(this.string);
    } else {
      output = output.write('[');
      Uri.writeHostLiteral(this.address, output);
      output = output.write(']');
    }
  }

  @Override
  public String toString() {
    if (this.string == null) {
      this.string = new StringBuilder(1 + this.address.length() + 1)
          .append('[').append(this.address).append(']').toString();
    }
    return this.string;
  }
}

File: swim-system-js/swim-core-js/@swim/uri/main/UriHostIPv6.ts
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Output} from "@swim/codec";
import {Uri} from "./Uri";
import {UriHost} from "./UriHost";

/** @hidden */
export class UriHostIPv6 extends UriHost {
  /** @hidden */
  constructor(address: string) {
    super();
    Object.defineProperty(this, "address", {
      value: address,
      enumerable: true,
    });
    Object.defineProperty(this, "stringValue", {
      value: void 0,
      enumerable: true,
      configurable: true,
    });
  }

  override readonly address!: string;

  override get ipv6(): string {
    return this.address;
  }

  override debug(output: Output): void {
    output = output.write("UriHost").write(46/*'.'*/).write("ipv6")
        .write(40/*'('*/).debug(this.address).write(41/*')'*/);
  }

  override display(output: Output): void {
    const stringValue = this.stringValue;
    if (stringValue !== void 0) {
      output = output.write(stringValue);
    } else {
      output = output.write(91/*'['*/);
      Uri.writeHostLiteral(this.address, output);
      output = output.write(93/*']'*/);
    }
  }

  /** @hidden */
  readonly stringValue!: string | undefined;

  override toString(): string {
    let stringValue = this.stringValue;
    if (stringValue === void 0) {
      stringValue = "[" + this.address + "]";
      Object.defineProperty(this, "stringValue", {
        value: stringValue,
        enumerable: true,
        configurable: true,
      });
    }
    return stringValue;
  }
}
